Search/ReIndex/Provider/DoctrineOrmProvider.php

```diff
@@ -0,0 +1,122 @@
+<?php
+
+/*
+ * This file is part of the MassiveSearchBundle
+ *
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Massive\Bundle\SearchBundle\Search\ReIndex\Provider;
+
+use Doctrine\ORM\EntityManagerInterface;
+use Massive\Bundle\SearchBundle\Search\ReIndex\ReIndexProviderInterface;
+use Metadata\MetadataFactory;
+
+class DoctrineOrmProvider implements ReIndexProviderInterface
+{
+    /**
+     * @var ClassMetadataFactory
+     */
+    private $entityManager;
+
+    /**
+     * @var MetadataFactory
+     */
+    private $searchMetadataFactory;
+
+    /**
+     * @deprecated BC hack. @see this::provide().
+     *
+     * @var object[]
+     */
+    private $cachedEntities = [];
+
+    public function __construct(
+        EntityManagerInterface $entityManager,
+        MetadataFactory $searchMetadataFactory
+    ) {
+        $this->entityManager = $entityManager;
+        $this->searchMetadataFactory = $searchMetadataFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getClassFqns()
+    {
+        $metadataFactory = $this->entityManager->getMetadataFactory();
+        $classFqns = [];
+
+        foreach ($metadataFactory->getAllMetadata() as $classMetadata) {
+            if (null === $this->searchMetadataFactory->getMetadataForClass($classMetadata->name)) {
+                continue;
+            }
+            $classFqns[] = $classMetadata->name;
+        }
+
+        return $classFqns;
+    }
+
+    /**
+     * BC Note: Previous versions of the MassiveSearchBundle expected a collection of entities to
+     *          be returned from the repository via. a custom repository method. The expected behavior
+     *          now is that a query builder will be passed and NOTHING should be returned.
+     *
+     * {@inheritdoc}
+     */
+    public function provide($classFqn, $offset, $maxResults)
+    {
+        if (!empty($this->cachedEntities)) {
+            return $this->sliceEntities($offset, $maxResults);
+        }
+
+        $repository = $this->entityManager->getRepository($classFqn);
+        $metadata = $this->searchMetadataFactory->getMetadataForClass($classFqn);
+
+        $repositoryMethod = $metadata->getOutsideClassMetadata()->getReindexRepositoryMethod();
+
+        $queryBuilder = $repository->createQueryBuilder('d');
+
+        if ($repositoryMethod) {
+            $result = $repository->$repositoryMethod($queryBuilder);
+
+            if (is_array($result)) {
+                @trigger_error('Reindex repository methods should NOT return anything. Use the passed query builder instead.');
```

Comment:
No this class is fine. The BC is just maintaining the old behavior
regarding returning entities.

On Mon, Feb 22, 2016 at 05:49:56AM -0800, Daniel Rotter wrote:
>    In [1]Search/ReIndex/Provider/DoctrineOrmProvider.php:
> 
>  > +        if (!empty($this->cachedEntities)) {
>  > +            return $this->sliceEntities($offset, $maxResults);
>  > +        }
>  > +
>  > +        $repository = $this->entityManager->getRepository($classFqn);
>  > +        $metadata = $this->searchMetadataFactory->getMetadataForClass($classFqn);
>  > +
>  > +        $repositoryMethod = $metadata->getOutsideClassMetadata()->getReindexRepositoryMethod();
>  > +
>  > +        $queryBuilder = $repository->createQueryBuilder('d');
>  > +
>  > +        if ($repositoryMethod) {
>  > +            $result = $repository->$repositoryMethod($queryBuilder);
>  > +
>  > +            if (is_array($result)) {
>  > +                @trigger_error('Reindex repository methods should NOT return anything. Use the passed query builder instead.');
> 
>    Is this class only here for keeping BC?
> 
>    —
>    Reply to this email directly or [2]view it on GitHub.
> 
>    Reverse link: [3]unknown
> 
> References
> 
>    Visible links
>    1. https://github.com/massiveart/MassiveSearchBundle/pull/77#discussion_r53626185
>    2. https://github.com/massiveart/MassiveSearchBundle/pull/77/files#r53626185
>    3. https://github.com/massiveart/MassiveSearchBundle/pull/77/files#r53626185